test/integration/framework/etcd.go

```diff
@@ -182,37 +182,58 @@ func EtcdMain(tests func() int) {
 	// Bail out early when -help was given as parameter.
 	flag.Parse()
 
-	before := runtime.NumGoroutine()
+	// Must be called *before* creating new goroutines.
+	goleakOpts := IgnoreBackgroundGoroutines()
+
+	goleakOpts = append(goleakOpts,
+		// lumberjack leaks a goroutine:
+		// https://github.com/natefinch/lumberjack/issues/56 This affects tests
+		// using --audit-log-path (like
+		// ./test/integration/apiserver/admissionwebhook/reinvocation_test.go).
+		// In normal production that should be harmless. We don't know here
+		// whether the test is using that, so we have to suppress reporting
+		// this leak for all tests.
+		//
+		// Both names occurred in practice.
+		goleak.IgnoreTopFunction("k8s.io/kubernetes/vendor/gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun"),
+		goleak.IgnoreTopFunction("gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun"),
+
+		// At least one test (test/integration/quota) starts an event
+		// recorder and that gets stuck in a loop
+		// (https://github.com/kubernetes/kubernetes/blob/ae23b0c11bc861bdca346666a739560ea3ee3145/staging/src/k8s.io/client-go/tools/record/event.go#L215-L231)
+		// which in the worst case loops twelve times with a delay of
+		// 10 seconds for each event before giving up.
+		//
+		// There's no good way to ignore this particular loop.
+		// Matching against Sleep is only indirect and not very
+		// precise.
+		goleak.IgnoreTopFunction("time.Sleep"),
+	)
+
 	stop, err := startEtcd()
 	if err != nil {
 		klog.Fatalf("cannot run integration tests: unable to start etcd: %v", err)
 	}
 	result := tests()
 	stop() // Don't defer this. See os.Exit documentation.
 
-	checkNumberOfGoroutines := func() (bool, error) {
-		// We leave some room for leaked goroutines as there are
-		// still some leaks, mostly:
-		// - leak from lumberjack package we're vendoring
-		// - leak from apiserve healthz
-		// - leak from opencensus library
-		// Once fixed, we should be able to bring it down to zero.
-		if dg := runtime.NumGoroutine() - before; dg <= 3 {
-			return true, nil
+	// Several tests don't wait for goroutines to stop. goleak.Find retries
+	// internally, but not long enough. 5 seconds seemed to be enough for
+	// most tests, even when testing in the CI.
+	timeout := 5 * time.Second
+	start := time.Now()
+	for {
+		if err := goleak.Find(goleakOpts...); err != nil {
+			if time.Now().Sub(start) >= timeout {
+				klog.ErrorS(err, "EtcdMain goroutine check")
+				result = 1
+				break
+			}
+		} else {
+			break
```

Comment:
one suggestion to reduce the indentation

```suggestion
		if err := goleak.Find(goleakOpts...); err == nil {
		        break
                 }
		if time.Now().Sub(start) >= timeout {
			klog.ErrorS(err, "EtcdMain goroutine check")
			result = 1
			break
		}
```

Reply:
That doesn't work as written, `err` won't be set after the `if`. Perhaps like this?

```
err := goleak.Find(goleakOpts...)
if err == nil {
	break
}
if time.Now().Sub(start) >= timeout {
	klog.ErrorS(err, "EtcdMain goroutine check")
	result = 1
	break
}
```

Reply:
oh, yeah, my bad, didn't see the error, in the klog, but you get what I mean